### PR TITLE
Openvpn fixes

### LINKED ISSUE #668: OpenVPN default cipher is not AES-CBC-256?
Hi Eric

I've done a couple of builds and much testing but the default is still blowfish?

see screen shot for defaults

default

### LINKED ISSUE #688: Panel: "OpenVPN Server: Allowed Clients" - regenerated VPN client does not block access to the server using previous key.
Panel: "OpenVPN Server: Allowed Clients" - regenerated VPN client does not block access to the server using previous key.

**Tested on:**

Router Model: Archer C7 V2 EU
Firmware: Gargoyle 1.10.0-ar71xx

**Test to reproduce:**

1. Create VPN client with name "test", push "Add" button, push "Save Changes". Download OVPN configuration and make sure your connection was established successfully.

2. Remove VPN client with name "test" and push "Save Changes".

3. Create VPN client with name "test" again, push "Add" button, push "Save Changes". And reboot router.

4. When router is up to use, try to connect to VPN using keys created first time. Make sure your connection was established successfully. **Furthermore, now you can use both OVPN configurations (first and last one) and they will work.**

Comparing new OVPN configuration file with OVPN created first time, i can see that **`<cert></cert>`** content differs, at: Validity, Modulus, "Subject Key Identifier", "Digital Signature" and CERTIFICATE. Also differs **`<key></key>`**.

All other content is the same, including **`<ca></ca>`** and **`<tls-auth></tls-auth>`**

**Old client should not be able to connect with his private/public key.**

### LINKED ISSUE #689: Panel: "OpenVPN Server: Allowed Clients" - checkbox "Enabled" for VPN client does not work.
Panel: "OpenVPN Server: Allowed Clients" - checkbox "Enabled" for VPN client does not work.

**Tested on:**

Router Model: Archer C7 V2 EU
Firmware: Gargoyle 1.10.0-ar71xx

**Test to reproduce:**
1. Create VPN client, push "Add" button, then push "SaveChanges".
2. Now, when VPN client is ready and checkbox "Enabled" is ticked, untick "Enabled" and push "SaveChanges".

**The result:** After changes apply, the checkbox is ticked again. And client still can connect.

**Disabled client should be not able to connect.**

### LINKED ISSUE #690: OpenVPN: iOS client error: route destinations other than vpn_gateway or net_gateway are not supported
**Tested on:**
Router Model: Archer C7 V2 EU
Firmware: Gargoyle 1.10.0-ar71xx

**Client:**
iOS 10.3.0
OpenVPN 1.1.1 build 212 (iOS 64-bit)

**Error:**
Error parsing IPv4 route: [route] [192.168.2.0] [255.255.255.0] [10.8.0.1] : tun_builder_route_error: route destinations other than vpn_gateway or net_gateway are not supported.

**Solution:**
Modify **/usr/lib/gargoyle/openvpn.sh** script lines:

From:
`echo "push \"route $subnet_ip $subnet_mask $openvpn_server_internal_ip\"" >> "$client_ccd_file"`
To:
`echo "push \"route $subnet_ip $subnet_mask vpn_gateway\"" >> "$client_ccd_file"`

From:
`echo "push \"route $subnet_ip $subnet_mask $openvpn_ip\"" >> "$client_ccd_file"`
To:
`echo "push \"route $subnet_ip $subnet_mask vpn_gateway\"" >> "$client_ccd_file"`

**To test solution:**
1. Rebuild routes using next commands:
```
source /usr/lib/gargoyle/openvpn.sh
update_routes
```
2. Reconnect with OpenVPN client.

### LINKED ISSUE #691: OpenVPN server route not being applied
Gargoyle 1.10.0

A client from behind the OpenVPN server can no longer reach a client behind the OpenVPN client.

Client information is no longer being pushed to the server routing table.

For full details see

https://www.gargoyle-router.com/phpbb/viewtopic.php?f=6&t=11339

image

### LINKED ISSUE #711: The setting of using another mac address is not remembered.
image
After I apply the settings, or when I open this settings page, the checkbox "use custom mac address" is always inactive. And if I apply the settings without re-setting it, the external mac address will return to the factory value.

### PR DESCRIPTION
**Pull request by lantis1008**
- Fix client disable checkbox not doing anything
- Implement verify-tls script to enable fast/easy temporary revocation of certificates
  
Client Verified
```
Tue Dec 19 20:57:55 2017 user.notice ovpn-cn-check: client1 verified OK
Tue Dec 19 20:57:55 2017 daemon.notice openvpn(custom_config)[26205]: 192.168.0.209:49488 VERIFY SCRIPT OK: depth=0, C=??, ST=UnknownProvince, L=UnknownCity, O=UnknownOrg, OU=UnknownOrgUnit, CN=client1, name=client1, emailAddress=client1@.com
```
Client Rejected
```
Tue Dec 19 20:54:08 2017 user.notice ovpn-cn-check: client1 not verified
Tue Dec 19 20:54:08 2017 daemon.warn openvpn(custom_config)[26205]: 192.168.0.209:33403 WARNING: Failed running command (--tls-verify script): external program exited with error status: 1
Tue Dec 19 20:54:08 2017 daemon.notice openvpn(custom_config)[26205]: 192.168.0.209:33403 VERIFY SCRIPT ERROR: depth=0, C=??, ST=UnknownProvince, L=UnknownCity, O=UnknownOrg, OU=UnknownOrgUnit, CN=client1, name=client1, emailAddress=client1@.com
```
Closes #689 

- Properly generate single .ovpn file (clean up the format)
- Allow individual clients to have their gateway address set to "vpn_gateway"

capture

Closes #690 

- Fix default cipher not changed to AES-265-CBC

Closes #668 

- Implement Certificate Revocation List (CRL) to allow us to permanently revoke certificates that are no longer required and therefore no longer valid
- Fix the random domain for email not being created properly. This makes the DN unique which makes revocation cleaner

Certificate Valid
```
Thu Jan  4 16:30:41 2018 daemon.notice openvpn(custom_config)[20897]: 192.168.0.1:61077 CRL CHECK OK: C=??, ST=UnknownProvince, L=UnknownCity, O=UnknownOrg, OU=UnknownOrgUnit, CN=client1, name=client1, emailAddress=client1@wcsvrlvajsjprxz.com
Thu Jan  4 16:30:41 2018 daemon.notice openvpn(custom_config)[20897]: 192.168.0.1:61077 VERIFY OK: depth=0, C=??, ST=UnknownProvince, L=UnknownCity, O=UnknownOrg, OU=UnknownOrgUnit, CN=client1, name=client1, emailAddress=client1@wcsvrlvajsjprxz.com
```
Certificate Revoked
```
Thu Jan  4 16:36:02 2018 daemon.notice openvpn(custom_config)[20897]: 192.168.0.1:65391 CRL CHECK FAILED: C=??, ST=UnknownProvince, L=UnknownCity, O=UnknownOrg, OU=UnknownOrgUnit, CN=client1, name=client1, emailAddress=client1@wcsvrlvajsjprxz.com (serial 01) is REVOKED
```
Closes #688 

- Fix unchecking of "Use Custom MAC Address" not actually returning the interfaces MAC to the default value

Closes #711 

- Fix Openvpn Status (openvpn_connections.sh) not hiding when server is disabled.
- Fix Openvpn not pushing new routes for the server by restarting the server when a client with a routed subnet is added/deleting. This will cause a momentary loss of connection for other clients, but they should all self re-establish. The reconnection period is equal to the keepalive setting (currently defaults to 180 seconds)

Closes #691

### CODE REVIEW

**Comment by lantis1008** [comment]
@ericpaulbishop i'm about to try and tackle #688 next

In order to do this, i'm going to want to revoke certificates (rather than just deleting them).
Have figured out how to do this [1], but i need an intact index file for the certificate database.
We are currently generating the certs/keys in /tmp/randomdir and then removing that dir afterwards, rather than operating at /usr/lib/easy-rsa. 
Is there a reason for this? Or is it purely a flash space conscious action?

It may not matter as i may be able to rescue the needed files and store them in /etc/openvpn with everything else and rebuild the folder as required (similar to what is already being done). This will be my next experiment based on your answer.

1. `Mon Dec 18 22:25:00 2017 daemon.notice openvpn(custom_config)[14401]: 192.168.0.209:56307 CRL CHECK FAILED: C=??, ST=UnknownProvince, L=UnknownCity, O=UnknownOrg, OU=UnknownOrgUnit, CN=client1, name=client1, emailAddress=client1@.com (serial 01) is REVOKED`

**Comment by ericpaulbishop** [comment]
Sorry for the delay on answering this: Writing to the /tmp/[random] dir is a flash space consideration and a consideration of the fact that reading/writing to a ramdisk (/tmp) is usually much faster than to flash. 

You've marked this patch set "NOT READY FOR MERGE", so when you feel it is ready for merge, please specifically let let me know here in the comments, otherwise I'll wait.

File: package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template [hunk 1]
```diff
@@ -1,20 +1,20 @@
-<div id="share_disk_container" class="row form-group">
```
[the remark this answers is not in the document]
Comment (lantis1008) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: I will check. I may be on outdated code there.

[the remark this answers is not in the document]
Comment (lantis1008) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: Before my change: 
image
Note the misalignment of the entire add shared disk section. I took this screenshot by taking the latest code from the repo... so you have something different going on locally if that isn't what you're seeing?

After my change:
image
Note the fixed alignment.

Alignment is all i'm trying to fix. I agree that zero margin between elements looks stupid, but my intention was to just make it a 30 second fix.

[the remark this answers is not in the document]
Comment (lantis1008) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: It looks to me like you removed "row" from div "sharing_add_controls_container" and added "indent". Am i on the right track?

[the remark this answers is not in the document]
Comment (lantis1008) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: happy to back the change out, but i don't like how it is now

[the remark this answers is not in the document]
Comment (d3fz) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: You're right. I must have forgotten to push the updated version of `usb_storage.sh` page to go with the template on that PR. Luckily, the solution is pretty minor and simple:

Just move `<%in templates/usb_storage_template %>` to inside the span right below (line 122)
```
<span class="col-xs-12">
    <%in templates/usb_storage_template %>
    <button id="add_share_button" class="btn btn-default" onclick="addNewShare()"><%~ ADsk %></button>
</span>
```

[the remark this answers is not in the document]
Comment (lantis1008) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: ok yes that is good solution. will you push that yourself or do you want me to? I will revert my commit soon

[the remark this answers is not in the document]
Comment (d3fz) [reply package/plugin-gargoyle-usb-storage/files/www/templates/usb_storage_template]: I'll handle it. Will probably add it among other fixes soon. Thanks for pointing it out.

**Comment by lantis1008** [comment]
Thought i was ready, need to take a look at #691 now

File: package/plugin-gargoyle-openvpn/files/www/js/openvpn.js [hunk 2]
```diff
@@ -214,12 +214,13 @@ function saveChanges()
...
 		// if anything in server section or client section has changed, restart openvpn
 		// otherwise we're just adding client certs to a server and restart shouldn't be needed
+		// unless we are adding/removing a client with a routed subnet, then we should give openvpn a kick to push the server routes
 		var openvpnCurrentlyRunning = (tunIp != "" && openvpnProc != "" && (remotePing != "" || openvpnConfig == "server"))
 		if(openvpnConfig == "disabled")
 		{
 			commands = "/etc/init.d/openvpn stop ; " + commands
 		}
-		else if(commands.match(/uci.*openvpn_gargoyle\.server\./) || openvpnConfig == "client" || (!openvpnCurrentlyRunning) )
+		else if(commands.match(/uci.*openvpn_gargoyle\.server\./) || commands.match(/uci.*openvpn_gargoyle\.[^server].*subnet_ip/) || openvpnConfig == "client" || (!openvpnCurrentlyRunning) )
```
Comment (lantis1008) [inline package/plugin-gargoyle-openvpn/files/www/js/openvpn.js]: although i now remark that perhaps the [^server] check is unnecessary and would fix the corner case... maybe fix when the complaints roll in

**Comment by lantis1008** [comment]
@ericpaulbishop 
This is now ready for merging. 
I've done my best to test Server configurations, but have done no client tests. As far as i can tell, clients should be unaffected.

Anyone else that wants to test please feel free and provide feedback. @ispyisail you seem to know a fair bit about openvpn, your input would be most appreciated.
